cmd/dep/ensure.go

```diff
@@ -262,76 +262,70 @@ func (s *stringSlice) Set(value string) error {
 	}
 
 	// try to split on '@'
-	var versionStr string
+	// When there is no `@`, use any version
+	versionStr := "*"
 	atIndex := strings.Index(arg, "@")
 	if atIndex > 0 {
 		parts := strings.SplitN(arg, "@", 2)
 		arg = parts[0]
 		versionStr = parts[1]
-		constraint.Constraint = deduceConstraint(parts[1])
 	}
-	// TODO: What if there is no @, assume default branch (which may not be master) ?
+
 	// TODO: if we decide to keep equals.....
 
 	// split on colon if there is a network location
+	var source string
 	colonIndex := strings.Index(arg, ":")
 	if colonIndex > 0 {
 		parts := strings.SplitN(arg, ":", 2)
 		arg = parts[0]
-		constraint.Ident.Source = parts[1]
+		source = parts[1]
 	}
 
 	pr, err := sm.DeduceProjectRoot(arg)
 	if err != nil {
-		return constraint, errors.Wrapf(err, "could not infer project root from dependency path: %s", arg) // this should go through to the user
+		return emptyPC, errors.Wrapf(err, "could not infer project root from dependency path: %s", arg) // this should go through to the user
 	}
 
 	if string(pr) != arg {
-		return constraint, errors.Errorf("dependency path %s is not a project root, try %s instead", arg, pr)
+		return emptyPC, errors.Errorf("dependency path %s is not a project root, try %s instead", arg, pr)
 	}
 
-	constraint.Ident.ProjectRoot = gps.ProjectRoot(arg)
-
-	// Below we are checking if the constraint we deduced was valid.
-	if v, ok := constraint.Constraint.(gps.Version); ok && versionStr != "" {
-		if v.Type() == gps.IsVersion {
-			// we hit the fall through case in deduce constraint, let's call out to network
-			// and get the package's versions
-			versions, err := sm.ListVersions(constraint.Ident)
-			if err != nil {
-				return constraint, errors.Wrapf(err, "list versions for %s", arg) // means repo does not exist
-			}
+	pi := gps.ProjectIdentifier{ProjectRoot: pr, Source: source}
+	c, err := deduceConstraint(versionStr, pi, sm)
+	if err != nil {
+		return emptyPC, err
+	}
+	return gps.ProjectConstraint{Ident: pi, Constraint: c}, nil
+}
 
-			var found bool
-			for _, version := range versions {
-				if versionStr == version.String() {
-					found = true
-					constraint.Constraint = version.Unpair()
-					break
-				}
-			}
+// deduceConstraint tries to puzzle out what kind of version is given in a string -
+// semver, a revision, or as a fallback, a plain tag
+func deduceConstraint(s string, pi gps.ProjectIdentifier, sm gps.SourceManager) (gps.Constraint, error) {
+	if s == "" {
+		// Find the default branch
+		// TODO (carolynvs): Should this be defined on SourceManager directly?
+		versions, err := sm.ListVersions(pi)
+		if err != nil {
+			return nil, errors.Wrapf(err, "list versions for %s(%s)", pi.ProjectRoot, pi.Source) // means repo does not exist
+		}
 
-			if !found {
-				return constraint, errors.Errorf("%s is not a valid version for the package %s", versionStr, arg)
+		for _, v := range versions {
+			if v.Type() == gps.IsBranch {
```

Comment:
We're not guaranteed sorted output from `ListVersions()` - the first branch we encounter might not be the default. Need to `gps.SortPairedForUpgrade()` on the slice to get the outcome we want.